### ISSUE TITLE
Allow configurable COMPOSE_FILE (use `.env` file)

### ISSUE THREAD

**Issue by mjbogusz** [microsoft/vscode-docker#3313]
It would be great if this extension allowed usage of alternative compose files.

`docker compose` already allows using files other than `docker-compose.yml` as its configuration - either via an argument (`--file`), via an environment variable (`COMPOSE_FILE`) or via a `.env` file.

In my case I have several compose files - one for local development, one for CI, one for production deployment etc, but I don't have a `docker-compose.yml` in the main project directory.
This isn't a problem as long as I launch everything from a terminal - my `.env` is being read and proper COMPOSE_FILE is used.
However this extension ignores the `.env` file and tries to use the default, non-existent `docker-compose.yml`.
My workaround so far was to softlink my preferred compose file to `docker-compose.yml` - but this is a workaround and it ignores other settings from this file.

Ideally this extension should use the `.env` file but also expose a "compose file" setting that would be passed into the `docker compose` commands via the `--file` argument thus overriding the one from `.env`.

**Comment by bwateratmsft** [comment]
In your scenario I'd use the `docker-compose` task: https://code.visualstudio.com/docs/containers/reference#_docker-compose-task

**Comment by mjbogusz** [comment]
Will this integrate with the docker sidebar and the `docker compose *` from command palette?

Also, I have other stuff in the `.env` file, like `COMPOSE_PROJECT_NAME` or, in this particular case, DB access data and Django's `SECRET_KEY` - how would I integrate that into the tasks solution?

**Comment by bwateratmsft** [comment]
The task allows you to specify `envFiles` so you can include your `.env` file in it. Unfortunately the `Compose Up` command and the right-click compose file -> compose up won't use the task--but you can run it by doing "Run Task" in the command palette and then choosing your task.

If you really want the palette command directly to Compose Up, you can use command customization to ensure the correct files get used: https://code.visualstudio.com/docs/containers/reference#_command-customization

**Comment by mjbogusz** [comment]
I'll try using the command customization, thanks for the tip!

I had a quick look at the sources of the extension and I can't find out why it doesn't use the `.env` file by default thought... `docker compose` should look for it in the directory it was launched and I assume it's the main workspace directory, right?

**Comment by bwateratmsft** [comment]
It depends whether you launch by right-clicking a compose file and doing Compose Up, or if you go through the command palette. In the former case, we will explicitly pass that file (and _only_ that file) via a `--file` argument. In the latter case, if there is only `docker-compose.yml` or `docker-compose.yml` + `docker-compose.override.yml`, we will run without any `--file` arguments, which should use the default behavior. If any other compose files are present we should prompt the user to choose which file to launch.

I'm not totally sure how the default behavior handles `.env`, and I also think that the new Compose V2 has some differences in that area as well.

**Comment by mjbogusz** [comment]
I've experimented a bit and apparently the extension expects a specific filename pattern for the compose files.

I.e. if my compose is named `docker-localhost.yml`, then running Compose Up only prompts:
> Couldn't find any docker-compose files in your workspace. Would you like to add Docker files to the workspace?

The same happens if I add the following to my `.code-workspace`:
```json
        "docker.commands.composeUp": [{
            "label": "Compose Up",
            "template": "docker compose -f docker-localhost.yml up ${detached} ${build}"
        }]
```
but if I rename that to `docker-compose.localhost.yml` then it gets picked up and started even without customizing the command.
I'm not sure why, as the `YAML_GLOB_PATTERN` template (`/src/constants.ts`) contains only the YAML extension.

**Comment by bwateratmsft** [comment]
Not all YAML files are compose files, hence why `YAML_GLOB_PATTERN` detects all YAML (it is being used for a different purpose than detecting compose files).

You can use a nonstandard filename like `docker-localhost.yml`, but you'll need to add the following to settings:

```json
    "files.associations": {
        "docker-localhost.yml": "dockercompose"
    }
```

This will enable right-clicking the file to do Compose Up. I did discover a bug where it will not show up in the Quick Pick list if you activate Compose Up from the palette--it gives the error you saw: "Couldn't find any docker-compose files in your workspace."

As a workaround I'd suggest renaming your file `docker-compose.localhost.yml`, IMO that's more in line with convention. We'll get that bug fixed in 1.19.0.

**Comment by bwateratmsft** [comment]
Fixed by #3323

**Comment by bwateratmsft** [comment]
This is now released in Docker extension version 1.19.0.
